### PR TITLE
🧪 Test USDT transaction payment

### PR DESCRIPTION
**Pull request by JuaniRios**
## What?
- Add an integration test to check the new USDT payment of transaction fees and tips.

## Testing?
`integration_tests::transaction_payment::fee_paid_with_foreign_assets`

### CODE REVIEW

**Comment by JuaniRios** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/Polimec/polimec-node/404?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#404** <a href="https://app.graphite.dev/github/pr/Polimec/polimec-node/404?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#403** <a href="https://app.graphite.dev/github/pr/Polimec/polimec-node/403?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#402** <a href="https://app.graphite.dev/github/pr/Polimec/polimec-node/402?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#401** <a href="https://app.graphite.dev/github/pr/Polimec/polimec-node/401?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#399** <a href="https://app.graphite.dev/github/pr/Polimec/polimec-node/399?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @JuaniRios and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: integration-tests/src/tests/transaction_payment.rs [hunk 1]
```diff
@@ -1,4 +1,181 @@
...
+use polimec_runtime::Header;
+use sp_arithmetic::{FixedPointNumber, FixedU128};
+use sp_core::H256;
+use sp_runtime::{
+	traits::{Dispatchable, Header as HeaderT, SignedExtension},
+	DigestItem,
+};
+use xcm_emulator::TestExt;
+generate_accounts!(ALICE, AUTHOR);
+use frame_support::traits::fungible::Inspect;
```
Comment (lrazovic) [inline integration-tests/src/tests/transaction_payment.rs:23]:
```suggestion
use frame_support::traits::fungible::Inspect;

generate_accounts!(ALICE, AUTHOR);
```

plus `cargo +nightly fmt --alll`

File: integration-tests/src/tests/transaction_payment.rs [hunk 2]
```diff
@@ -1,4 +1,181 @@
...
+			&polimec_runtime::BlockchainOperationTreasury::get(),
+			100 * usdt_unit,
+		);
+		// Block author doesn't need to have any balance, as the tip is bigger than ED.
+
+		let paid_call = PolimecCall::System(frame_system::Call::remark { remark: vec![69, 69] });
+
+		type TxPaymentExtension = pallet_asset_tx_payment::ChargeAssetTxPayment<PolimecRuntime>;
+		let signed_extension =
+			pallet_asset_tx_payment::ChargeAssetTxPayment::<PolimecRuntime>::from(10 * plmc_unit, Some(usdt_id));
```
Comment (lrazovic) [inline integration-tests/src/tests/transaction_payment.rs:82]: Can you please clarify with a code comment what's happening on this line? Is it correct to say that "a user wants to leave as a tip the equivalent in USDT of 10 PLMC"?

File: integration-tests/src/tests/transaction_payment.rs [hunk 3]
```diff
@@ -1,4 +1,181 @@
...
+
+		let dispatch_info = paid_call.get_dispatch_info();
+		let FeeDetails { inclusion_fee, tip } =
+			polimec_runtime::TransactionPayment::compute_fee_details(0u32, &dispatch_info, 10u128 * plmc_unit);
```
Comment (lrazovic) [inline integration-tests/src/tests/transaction_payment.rs:86]: **Q:** Is it ok to set the `len` to 0? Wouldn't be more precise to use the `len` of the  `dispatch_info`?

File: integration-tests/src/tests/transaction_payment.rs [hunk 4]
```diff
@@ -1,4 +1,181 @@
...
+			PolimecForeignAssets::balance(usdt_id, polimec_runtime::BlockchainOperationTreasury::get());
+		let prev_blockchain_operation_treasury_plmc_balance =
+			PolimecBalances::balance(&polimec_runtime::BlockchainOperationTreasury::get());
+		let prev_block_author_usdt_balance = PolimecForeignAssets::balance(usdt_id, block_author.clone());
+		let prev_block_author_plmc_balance = PolimecBalances::balance(&block_author.clone());
+
+		let pre = signed_extension.pre_dispatch(&alice, &paid_call, &dispatch_info, 8).unwrap();
+		let post_info = paid_call.clone().dispatch(PolimecOrigin::signed(alice.clone())).expect("call dispatch failed");
+
+		TxPaymentExtension::post_dispatch(Some(pre), &dispatch_info, &post_info, 0usize, &Ok(())).unwrap();
```
Comment (lrazovic) [inline integration-tests/src/tests/transaction_payment.rs:110]:
```suggestion
// Executes the `pre_dispatch` check for the transaction using the signed extension.
// Dispatches the `paid_call` signed by `alice` and ensures it executes successfully.
// Runs the `post_dispatch` logic to verify correct post-transaction handling, 
//    including fee calculation and cleanup using the `TxPaymentExtension`.
		let pre = signed_extension.pre_dispatch(&alice, &paid_call, &dispatch_info, 8).unwrap();
		let post_info = paid_call.clone().dispatch(PolimecOrigin::signed(alice.clone())).expect("call dispatch failed");
		TxPaymentExtension::post_dispatch(Some(pre), &dispatch_info, &post_info, 0usize, &Ok(())).unwrap();
```

File: integration-tests/src/tests/transaction_payment.rs [hunk 5]
```diff
@@ -1,4 +1,181 @@
...
+		let prev_alice_usdt_balance = PolimecForeignAssets::balance(usdt_id, alice.clone());
+		let prev_alice_plmc_balance = PolimecBalances::balance(&alice);
+		let prev_blockchain_operation_treasury_usdt_balance =
+			PolimecForeignAssets::balance(usdt_id, polimec_runtime::BlockchainOperationTreasury::get());
+		let prev_blockchain_operation_treasury_plmc_balance =
+			PolimecBalances::balance(&polimec_runtime::BlockchainOperationTreasury::get());
+		let prev_block_author_usdt_balance = PolimecForeignAssets::balance(usdt_id, block_author.clone());
+		let prev_block_author_plmc_balance = PolimecBalances::balance(&block_author.clone());
+
+		let pre = signed_extension.pre_dispatch(&alice, &paid_call, &dispatch_info, 8).unwrap();
```
Comment (lrazovic) [inline integration-tests/src/tests/transaction_payment.rs:107]: **Q:** Why is the `len` 8 here?

File: integration-tests/src/tests/transaction_payment.rs (hunk 4 above)
Comment (lrazovic) [inline integration-tests/src/tests/transaction_payment.rs:110]: Q: Why is the `len` 0 here, if in the `pre_dispatch` is set to 0?

File: integration-tests/src/tests/transaction_payment.rs [hunk 6]
```diff
@@ -1,4 +1,181 @@
...
+		let post_alice_plmc_balance = PolimecBalances::balance(&alice);
+		let post_blockchain_operation_treasury_usdt_balance =
+			PolimecForeignAssets::balance(usdt_id, polimec_runtime::BlockchainOperationTreasury::get());
+		let post_blockchain_operation_treasury_plmc_balance =
+			PolimecBalances::balance(&polimec_runtime::BlockchainOperationTreasury::get());
+		let post_block_author_usdt_balance = PolimecForeignAssets::balance(usdt_id, block_author.clone());
+		let post_block_author_plmc_balance = PolimecBalances::balance(&block_author.clone());
+
+		assert_eq!(prev_alice_usdt_balance - post_alice_usdt_balance, expected_usd_fee + expected_usd_tip);
+		assert_eq!(post_alice_plmc_balance - prev_alice_plmc_balance, 0u128);
```
Comment (lrazovic) [inline integration-tests/src/tests/transaction_payment.rs:122]:
```suggestion
		assert_eq!(post_alice_plmc_balance, prev_alice_plmc_balance);
```

File: integration-tests/src/tests/transaction_payment.rs [hunk 7]
```diff
@@ -1,4 +1,181 @@
...
+		assert_eq!(
+			post_blockchain_operation_treasury_usdt_balance - prev_blockchain_operation_treasury_usdt_balance,
+			expected_usd_fee
+		);
+		assert_eq!(
+			post_blockchain_operation_treasury_plmc_balance - prev_blockchain_operation_treasury_plmc_balance,
+			0u128
+		);
```
Comment (lrazovic) [inline integration-tests/src/tests/transaction_payment.rs:130]:
```suggestion
		assert_eq!(
			post_blockchain_operation_treasury_plmc_balance, prev_blockchain_operation_treasury_plmc_balance,
		);
```

File: integration-tests/src/tests/transaction_payment.rs [hunk 8]
```diff
@@ -1,4 +1,181 @@
...
+		assert_eq!(post_block_author_usdt_balance - prev_block_author_usdt_balance, expected_usd_tip);
+		assert_eq!(post_block_author_plmc_balance - prev_block_author_plmc_balance, 0u128);
```
Comment (lrazovic) [inline integration-tests/src/tests/transaction_payment.rs:132]:
```suggestion
		assert_eq!(post_block_author_plmc_balance, prev_block_author_plmc_balance);
```

File: integration-tests/src/tests/transaction_payment.rs [hunk 9]
```diff
@@ -1,4 +1,181 @@
...
+		let post_alice_usdt_balance = PolimecForeignAssets::balance(usdt_id, alice.clone());
+		let post_alice_plmc_balance = PolimecBalances::balance(&alice);
+		let post_blockchain_operation_treasury_usdt_balance =
+			PolimecForeignAssets::balance(usdt_id, polimec_runtime::BlockchainOperationTreasury::get());
+		let post_blockchain_operation_treasury_plmc_balance =
+			PolimecBalances::balance(&polimec_runtime::BlockchainOperationTreasury::get());
+		let post_block_author_usdt_balance = PolimecForeignAssets::balance(usdt_id, block_author.clone());
+		let post_block_author_plmc_balance = PolimecBalances::balance(&block_author.clone());
+
+		assert_eq!(post_alice_usdt_balance - prev_alice_usdt_balance, 0u128);
```
Comment (lrazovic) [inline integration-tests/src/tests/transaction_payment.rs:168]: same as above

File: runtimes/shared-configuration/src/fee.rs [hunk 10]
```diff
@@ -242,7 +242,7 @@ where
 		// Calculate how much refund we should return.
 		let (final_fee, refund) = paid.split(converted_fee);
 		// Split the tip from the fee
-		let (final_fee_minus_tip, final_tip) = final_fee.split(converted_tip);
+		let (final_tip, final_fee_minus_tip) = final_fee.split(converted_tip);
```
Comment (lrazovic) [inline runtimes/shared-configuration/src/fee.rs:245]: **Q:** Is this change intended?

**Review by lrazovic — CHANGES_REQUESTED** [review]
Just a few questions around the `dispatch` len, the logic looks good.

File: runtimes/shared-configuration/src/fee.rs (hunk 10 above)
Comment (JuaniRios) [reply runtimes/shared-configuration/src/fee.rs:245]: Yes, when testing I realised I inverted the fields. This change should have been absorbed on the PR below, but too late now :c

File: integration-tests/src/tests/transaction_payment.rs (hunk 1 above)
Comment (JuaniRios) [reply integration-tests/src/tests/transaction_payment.rs:23]: doing fmt changes might fuck up the PRs on top. Let's make sure teh fmt check passes on the final PR before merging the stack

File: integration-tests/src/tests/transaction_payment.rs (hunk 3 above)
Comment (JuaniRios) [reply integration-tests/src/tests/transaction_payment.rs:86]: I couldn't find any info on what len means, and saw Just was using 0 on the dispenser tests. IDK

File: integration-tests/src/tests/transaction_payment.rs (hunk 4 above)
Comment (JuaniRios) [reply integration-tests/src/tests/transaction_payment.rs:110]: Added in my own PR with other changes

File: integration-tests/src/tests/transaction_payment.rs (hunk 5 above)
Comment (JuaniRios) [reply integration-tests/src/tests/transaction_payment.rs:107]: Not sure, I think it was failing with 0. Ill investigate lengths further

File: integration-tests/src/tests/transaction_payment.rs (hunk 3 above)
Comment (JuaniRios) [reply integration-tests/src/tests/transaction_payment.rs:89]: Saw some other tests where they put the encoded call len in that argument, so I did that on these tests.

File: integration-tests/src/tests/transaction_payment.rs (hunk 4 above)
Comment (JuaniRios) [reply integration-tests/src/tests/transaction_payment.rs:117]: Saw some other tests where they put the encoded call len in that argument, so I did that on these tests.

**Review by lrazovic — APPROVED** [review]

**Comment by JuaniRios** [comment]
### Merge activity

* **Oct 10, 10:08 AM EDT**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/Polimec/polimec-node/404).
